crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/WebCrawler.java

```diff
@@ -405,184 +383,187 @@ public void visit(Page page) {
+				
+				handlePageStatusCode(curURL, statusCode,
+						EnglishReasonPhraseCatalog.INSTANCE.getReason(statusCode, Locale.ENGLISH));
+				
+				// Finds the status reason for all known statuses
+				
+				page.setFetchResponseHeaders(fetchResult.getResponseHeaders());
+				page.setStatusCode(statusCode);
+				
+				
+				if (statusCode < HttpStatus.SC_SUCCESS || statusCode > 299) { // Not 2XX: 2XX status codes indicate success
+					if (statusCode == HttpStatus.SC_MOVED_PERMANENTLY //
+							|| statusCode == HttpStatus.SC_MOVED_TEMPORARILY //
+							|| statusCode == HttpStatus.SC_MULTIPLE_CHOICES //
+							|| statusCode == HttpStatus.SC_SEE_OTHER //
+							|| statusCode == HttpStatus.SC_TEMPORARY_REDIRECT //
+							|| statusCode == HttpStatus.SC_PERMANENT_REDIRECT //
+					) { // is 3xx  todo
+						// follow https://issues.apache.org/jira/browse/HTTPCORE-389
+						return processFetchResultRedirect(fetchResult, page);
+					} else { // All other http codes other than 3xx & 200
+						String description = EnglishReasonPhraseCatalog.INSTANCE.getReason(//
+								fetchResult.getStatusCode(), Locale.ENGLISH); // Finds the status reason for all known statuses
+						String contentType = fetchResult.getEntity() == null ? "" //
+								: fetchResult.getEntity().getContentType() == null ? "" //
+										: fetchResult.getEntity().getContentType();
+						onUnexpectedStatusCode(curURL.getURL(), fetchResult.getStatusCode(), contentType, description);
+						return false;
+					}
+				} else { // if status code is 200
+					return handleFetchResultSuccess(fetchResult, page);
+				}
+			} catch (PageBiggerThanMaxSizeException e) {
+				onPageBiggerThanMaxSize(curURL.getURL(), e.getPageSize());
+			} catch (ContentFetchException | SocketTimeoutException cfe) {
+				onContentFetchError(curURL);
+			} catch (Exception e) {
+				onUnhandledException(curURL, e);
+			}
+			return false;
+		}
+
+		private boolean handleFetchResultSuccess(final PageFetchResult fetchResult, final Page page)
+				throws IOException,
+				ContentFetchException,
+				InterruptedException
+		{
+			final WebURL curURL = page.getWebURL();
+			
+			if (!curURL.getURL().equals(fetchResult.getFetchedUrl())) {
+				if (docIdServer.isSeenBefore(fetchResult.getFetchedUrl())) {
+					logger.debug("Redirect page: {} has already been seen", curURL);
+					return true;
+				}
+				curURL.setURL(fetchResult.getFetchedUrl());
+				curURL.setDocid(docIdServer.getNewDocID(fetchResult.getFetchedUrl()));
+			}
+			
+			if (!fetchResult.fetchContent(page, myController.getConfig().getMaxDownloadSize())) {
+				throw new ContentFetchException();
+			}
+			
+			if (page.isTruncated()) {
+				logger.warn("Warning: unknown page size exceeded max-download-size, truncated to: ({}), at URL: {}",
+						myController.getConfig().getMaxDownloadSize(), curURL.getURL());
+			}
+			
+			try {
+				// TODO Passing content from first parameter as second parameter is confusing.
```

Comment:
Well it actually starts inside:
```java
private boolean processPage(WebURL curURL) {
```
There the page-object is being constructed and populated:
```java
Page page = new Page(curURL);
```
I broke up the processPage()-method into 3 sub-methods (success, some specific redirect statuses, and the rest), so the fact that this curURL is passed separately became very obvious, and I took it out of the argument list. This passing of the page object + curURL separately ripples through to here (the TODO), but changing the argument list of the parse()-method would probably be a big API change as I expect it to have been there nearly since the beginning.

At first I was wondering if they could ever be different (the url of the page object or the "current url"), but that wouldn't make sense, so it looked less confusing to use the page-object as central point to get to that information.